server/datastore/mysql/hosts.go

```diff
@@ -1926,72 +1953,92 @@ WHERE
-    GROUP BY
-      h.os_version,
-      h.platform,
-      h.team_id
-  ) as team_os_versions
-GROUP BY
-  team_id
-UNION
-SELECT
-  t.id,
-  'os_versions' type,
-  JSON_ARRAY() json_value
-FROM
-  teams t
-WHERE NOT EXISTS (
-  SELECT
-    id
-  FROM
-    hosts h
-  WHERE
-    t.id = h.team_id
-)
-ON DUPLICATE KEY UPDATE
-  json_value = VALUES(json_value),
-  updated_at = CURRENT_TIMESTAMP
-`
+	selectStmt := `
+	SELECT 
+		COUNT(*) hosts_count,  
+		h.team_id, 
+		os.id,
+		os.name, 
+		os.version, 
+		os.platform
+	FROM hosts h
+	JOIN host_operating_system hos ON h.id = hos.host_id 
+	JOIN operating_systems os ON hos.os_id = os.id 
+	GROUP BY team_id, os_id
+	`
 
-	_, err := ds.writer.ExecContext(ctx, sql)
-	if err != nil {
-		return ctxerr.Wrapf(ctx, err, "update aggregated stats for os versions")
+	var rows []struct {
+		HostsCount int    `db:"hosts_count"`
+		Name       string `db:"name"`
+		Version    string `db:"version"`
+		Platform   string `db:"platform"`
+		ID         uint   `db:"id"`
+		TeamID     *uint  `db:"team_id"`
+	}
+	if err := sqlx.SelectContext(ctx, ds.reader, &rows, selectStmt); err != nil {
+		return ctxerr.Wrap(ctx, err, "update os versions")
+	}
+
+	// each team has a slice of stats with team host counts per os version
+	statsByTeamID := make(map[uint][]fleet.OSVersion)
+	// stats are also aggregated globally per os version
+	globalStats := make(map[uint]fleet.OSVersion)
+
+	for _, r := range rows {
+		os := fleet.OSVersion{
+			HostsCount: r.HostsCount,
+			Name:       fmt.Sprintf("%s %s", r.Name, r.Version),
+			NameOnly:   r.Name,
+			Version:    r.Version,
+			Platform:   r.Platform,
+			ID:         r.ID,
+		}
+		// increment global stats
+		if _, ok := globalStats[os.ID]; !ok {
+			globalStats[os.ID] = os
+		} else {
+			newStats := globalStats[os.ID]
+			newStats.HostsCount += r.HostsCount
+			globalStats[os.ID] = newStats
+		}
+		// push to team stats if applicable
+		if r.TeamID != nil {
+			statsByTeamID[*r.TeamID] = append(statsByTeamID[*r.TeamID], os)
+		}
+	}
+
+	// if an existing team has no hosts assigned, we still want to store empty stats
+	var teamIDs []uint
+	if err := sqlx.SelectContext(ctx, ds.reader, &teamIDs, "SELECT id FROM teams"); err != nil {
+		return ctxerr.Wrap(ctx, err, "update os versions")
+	}
+	for _, id := range teamIDs {
+		if _, ok := statsByTeamID[id]; !ok {
+			statsByTeamID[id] = []fleet.OSVersion{}
+		}
+	}
+
+	// global stats are stored under id 0
+	for _, os := range globalStats {
+		statsByTeamID[0] = append(statsByTeamID[0], os)
+	}
+
+	// assemble values as arguments for insert statement
+	args := make([]interface{}, 0, len(statsByTeamID)*3)
+	for id, stats := range statsByTeamID {
+		jsonValue, err := json.Marshal(stats)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "marshal os version stats")
+		}
+		args = append(args, id, "os_versions", jsonValue)
+	}
+
+	insertStmt := "INSERT INTO aggregated_stats (id, type, json_value) VALUES "
```

Comment:
We are missing some tests with "empty tables" (probably with no `teams` and/or `hosts`?):
![Screen Shot 2022-08-12 at 08 18 36](https://user-images.githubusercontent.com/2073526/184344076-db44a8c4-be35-4181-8aa4-3ed2e5ec33e0.png)